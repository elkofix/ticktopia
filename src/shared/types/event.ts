export interface User {
  id: string;
  email: string;
  name: string;
  lastname: string;
  isActive: boolean;
  roles: Role[];
}

export interface Event {
  id: string;
  name: string;
  bannerPhotoUrl: string;
  isPublic: boolean;
  user: User;
}

export interface GetEventsParams {
  limit?: number;
  offset?: number;
}
