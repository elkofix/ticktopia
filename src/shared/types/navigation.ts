export interface NavigationItem {
    label: string
    href: string
    icon: React.ElementType
    roles: Role[]
    priority?: number 
}
