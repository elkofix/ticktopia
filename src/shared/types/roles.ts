type Role = "admin" | "client" | "event-manager" | "ticketChecker"