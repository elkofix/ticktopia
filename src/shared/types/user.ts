export interface AuthUser {
    id: string;
    email: string;
    isActive: boolean;
    roles: string[];
}
export interface AuthState {
  user: AuthUser | null;
  isAuthenticated: boolean;
  isLoading: boolean;
  error: string | null;
}