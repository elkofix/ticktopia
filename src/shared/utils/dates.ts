export function formateDate(date: string) {
    return new Date(date).toLocaleDateString('es-ES', {
        day: 'numeric',
        month: 'long',
        year: 'numeric',
        hour: '2-digit',
        minute: '2-digit'
    });
}
export const formatDateYMD = (dateString: string) => {
    const date = new Date(dateString);
    return date.toLocaleDateString('es-ES', {
        year: 'numeric',
        month: 'long',
        day: 'numeric'
    });
};

export const getDaysBetween = (date: string, date2: string) => {
    const availabilityDate = new Date(date);
    const eventOpenDate = new Date(date2);
    const diffTime = eventOpenDate.getTime() - availabilityDate.getTime();
    const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
    return diffDays;
};


export const isLaterDate = (date: string) => {
    const now = new Date();
    const availabilityDate = new Date(date);
    return now >= availabilityDate;
};
