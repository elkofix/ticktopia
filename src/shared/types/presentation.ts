import { Event } from "./event";

export interface Presentation {
  idPresentation: string;
  place: string;
  event: Event;
  capacity: number;
  price: number;
  openDate: string;
  startDate: string;
  latitude: number;
  longitude: number;
  description: string;
  ticketAvailabilityDate: string;
  ticketSaleAvailabilityDate: string;
  city: string;
}