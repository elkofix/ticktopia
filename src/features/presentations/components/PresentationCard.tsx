import Image from 'next/image';
import Link from 'next/link';
import { Presentation } from '@/shared/types/presentation';
import { formateDate } from '@/shared/utils/dates';

interface PresentationCardProps {
  presentation: Presentation;
  bannerPhotoUrl: string;
  name: string;
}

export default function PresentationCard({ presentation, bannerPhotoUrl, name }: PresentationCardProps) {
  return (
    <div
      className="group flex flex-col md:flex-row bg-gray-50 rounded-lg overflow-hidden hover:shadow-md transition-all duration-200 hover:bg-gray-100"
      data-testid="presentation-card"
    >
      {/* Presentation Banner */}
      <div
        className="relative w-full h-40 md:w-48 md:h-32 lg:w-56 lg:h-36 flex-shrink-0"
        data-testid="image-container"
      >
        <Image
          src={bannerPhotoUrl}
          alt={name}
          className="w-full h-full object-cover group-hover:scale-105 transition-transform duration-300"
          fill
          sizes="(max-width: 768px) 100vw, (max-width: 1024px) 200px, 250px"
          onError={(e) => {
            e.currentTarget.src = '/placeholder-presentation.jpg';
          }}
          data-testid="presentation-image"
        />
      </div>

      {/* Presentation Info */}
      <div className="flex-1 p-4 md:p-6 flex flex-col lg:flex-row justify-between gap-4">
        <div className="flex-1 min-w-0">
          <h3 className="text-lg md:text-xl font-bold text-gray-900 mb-2 md:mb-3 group-hover:text-brand transition-colors line-clamp-1">
            {name}
          </h3>

          <div className="space-y-2" >
            <div className="flex items-center text-gray-600">
              <svg className="w-4 h-4 mr-2 text-brand flex-shrink-0" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M8 7V3a1 1 0 011-1h6a1 1 0 011 1v4m3 0V6a2 2 0 00-2-2H7a2 2 0 00-2 2v1m16 0v12a2 2 0 01-2 2H5a2 2 0 01-2-2V7h16z" />
              </svg>
              <span className="text-sm font-medium line-clamp-1" data-testid="presentation-date">{formateDate(presentation.startDate)}</span>
            </div>

            <div className="flex items-center text-gray-600">
              <svg className="w-4 h-4 mr-2 text-brand flex-shrink-0" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M17.657 16.657L13.414 20.9a1.998 1.998 0 01-2.827 0l-4.244-4.243a8 8 0 1111.314 0z" />
                <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M15 11a3 3 0 11-6 0 3 3 0 016 0z" />
              </svg>
              <span className="text-sm line-clamp-1">
                <span className="text-gray-400 mx-1 hidden md:inline">•</span>
                <span data-testid="presentation-location">{presentation.city}</span>
              </span>
            </div>

            {presentation.description && (
              <p
                className="text-sm text-gray-500 mt-2 line-clamp-2"
                data-testid="presentation-description"
              >
                {presentation.description}
              </p>
            )}
          </div>
        </div>

        {/* Action Button - Ajustado para no cortarse */}
        <div className="flex-shrink-0 self-end md:self-center">
          <Link href={`/presentation/${presentation.idPresentation}`} className="inline-block min-w-[150px]"  data-testid="presentation-link">
            <button className="bg-brand text-white px-4 py-2 rounded-lg font-medium hover:bg-brand-dark transition-all duration-200 transform hover:scale-105 shadow-sm hover:shadow-md w-full text-center">
              Ver Presentación
            </button>
          </Link>
        </div>
      </div>
    </div>
  );
}