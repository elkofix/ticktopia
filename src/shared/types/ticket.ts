import { Presentation } from "./presentation";
import { AuthUser } from "./user";

export interface Ticket {
    id: string;
    buyDate: string;
    isRedeemed: boolean;
    isActive: boolean;
    quantity: number;
    user: AuthUser;
    presentation: Presentation

}