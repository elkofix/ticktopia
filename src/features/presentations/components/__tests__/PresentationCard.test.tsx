import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import PresentationCard from '../PresentationCard';
import { formateDate } from '../../../../shared/utils/dates';
import '@testing-library/jest-dom';
import { Presentation } from '@/shared/types/presentation';

// Mock next/image and next/link
jest.mock('next/image', () => {
  const MockNextImage = (props: any) => {
    // eslint-disable-next-line jsx-a11y/alt-text
    return <img {...props} />;
  };
  MockNextImage.displayName = 'MockNextImage';

  return {
    __esModule: true,
    default: MockNextImage,
  };
});


jest.mock('next/link', () => {
  const MockLink = ({ children, href }: { children: React.ReactNode; href: string }) => (
    <a href={href}>{children}</a>
  );
  MockLink.displayName = 'MockNextLink';
  return MockLink;
});

// Mock the date formatter
jest.mock('../../../../shared/utils/dates', () => ({
  formateDate: jest.fn((date) => `Formatted: ${date}`),
}));

describe('PresentationCard', () => {
  const mockPresentation: Presentation = {
    idPresentation: 'pres-123',
    place: 'Main Hall',
    event: {
      id: 'event-123',
      name: 'Test Event',
      bannerPhotoUrl: '/event-banner.jpg',
      isPublic: true,
      user: {
        id: 'user-123',
        email: 'organizer@test.com',
        name: 'Organizer',
        lastname: 'Test',
        isActive: true,
        roles: ['event-manager']
      }
    },
    capacity: 100,
    price: 50,
    openDate: '2023-01-01T18:00:00Z',
    startDate: '2023-01-01T20:00:00Z',
    latitude: 40.7128,
    longitude: -74.0060,
    description: 'This is a test presentation description that might be long.',
    ticketAvailabilityDate: '2022-12-01T00:00:00Z',
    ticketSaleAvailabilityDate: '2022-11-01T00:00:00Z',
    city: 'New York'
  };

  const mockProps = {
    presentation: mockPresentation,
    bannerPhotoUrl: '/banner.jpg',
    name: 'Test Presentation'
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with all props', () => {
    render(<PresentationCard {...mockProps} />);

    // Check container
    const container = screen.getByTestId('presentation-card');
    expect(container).toBeInTheDocument();
    expect(container).toHaveClass('group');
    expect(container).toHaveClass('flex-col');
    expect(container).toHaveClass('md:flex-row');

    // Check image
    const image = screen.getByTestId('presentation-image');
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src', mockProps.bannerPhotoUrl);
    expect(image).toHaveAttribute('alt', mockProps.name);

    // Check title
    expect(screen.getByText(mockProps.name)).toBeInTheDocument();
  });

  it('handles image error', () => {
    render(<PresentationCard {...mockProps} />);
    
    const image = screen.getByTestId('presentation-image');
    fireEvent.error(image);
    
    expect(image).toHaveAttribute('src', '/placeholder-presentation.jpg');
  });

  it('renders without description when not provided', () => {
    const presentationWithoutDesc = {
      ...mockPresentation,
      description: ''
    };
    
    render(<PresentationCard {...mockProps} presentation={presentationWithoutDesc} />);
    
    expect(screen.queryByTestId('presentation-description')).not.toBeInTheDocument();
  });

  it('applies hover styles', () => {
    render(<PresentationCard {...mockProps} />);
    
    const container = screen.getByTestId('presentation-card');
    expect(container).toHaveClass('hover:shadow-md');
    expect(container).toHaveClass('hover:bg-gray-100');
    
    const image = screen.getByTestId('presentation-image');
    expect(image).toHaveClass('group-hover:scale-105');
    
    const title = screen.getByText(mockProps.name);
    expect(title).toHaveClass('group-hover:text-brand');
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<PresentationCard {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('has correct responsive classes', () => {
    render(<PresentationCard {...mockProps} />);
    
    const container = screen.getByTestId('presentation-card');
    expect(container).toHaveClass('flex-col');
    expect(container).toHaveClass('md:flex-row');
    
    const imageContainer = screen.getByTestId('image-container');
    expect(imageContainer).toHaveClass('w-full');
    expect(imageContainer).toHaveClass('md:w-48');
    expect(imageContainer).toHaveClass('lg:w-56');
  });
});