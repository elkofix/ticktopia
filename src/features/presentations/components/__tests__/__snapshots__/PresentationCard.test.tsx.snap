// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PresentationCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="group flex flex-col md:flex-row bg-gray-50 rounded-lg overflow-hidden hover:shadow-md transition-all duration-200 hover:bg-gray-100"
    data-testid="presentation-card"
  >
    <div
      class="relative w-full h-40 md:w-48 md:h-32 lg:w-56 lg:h-36 flex-shrink-0"
      data-testid="image-container"
    >
      <img
        alt="Test Presentation"
        class="w-full h-full object-cover group-hover:scale-105 transition-transform duration-300"
        data-testid="presentation-image"
        sizes="(max-width: 768px) 100vw, (max-width: 1024px) 200px, 250px"
        src="/banner.jpg"
        style="position: absolute; height: 100%; width: 100%; object-fit: cover;"
      />
    </div>
    <div
      class="flex-1 p-4 md:p-6 flex flex-col lg:flex-row justify-between gap-4"
    >
      <div
        class="flex-1 min-w-0"
      >
        <h3
          class="text-lg md:text-xl font-bold text-gray-900 mb-2 md:mb-3 group-hover:text-brand transition-colors line-clamp-1"
        >
          Test Presentation
        </h3>
        <div
          class="space-y-2"
        >
          <div
            class="flex items-center text-gray-600"
          >
            <svg
              class="w-4 h-4 mr-2 text-brand flex-shrink-0"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
            >
              <path
                d="M8 7V3a1 1 0 011-1h6a1 1 0 011 1v4m3 0V6a2 2 0 00-2-2H7a2 2 0 00-2 2v1m16 0v12a2 2 0 01-2 2H5a2 2 0 01-2-2V7h16z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <span
              class="text-sm font-medium line-clamp-1"
              data-testid="presentation-date"
            >
              Formatted: 2023-01-01T20:00:00Z
            </span>
          </div>
          <div
            class="flex items-center text-gray-600"
          >
            <svg
              class="w-4 h-4 mr-2 text-brand flex-shrink-0"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
            >
              <path
                d="M17.657 16.657L13.414 20.9a1.998 1.998 0 01-2.827 0l-4.244-4.243a8 8 0 1111.314 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
              <path
                d="M15 11a3 3 0 11-6 0 3 3 0 016 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <span
              class="text-sm line-clamp-1"
            >
              <span
                class="text-gray-400 mx-1 hidden md:inline"
              >
                â€¢
              </span>
              <span
                data-testid="presentation-location"
              >
                New York
              </span>
            </span>
          </div>
          <p
            class="text-sm text-gray-500 mt-2 line-clamp-2"
            data-testid="presentation-description"
          >
            This is a test presentation description that might be long.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
